ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java

```diff
@@ -158,19 +197,61 @@ private void removeAppliedRangeActions(Set<RangeAction> rangeActions) {
     /**
      * Removes from a set of RangeActions the ones with the biggest priority or sensitivity on the most limiting element
      *
-     * @param rangeActions the set of RangeActions to filter
-     * @param numberToRemove       the number of RangeActions to remove from the set
+     * @param rangeActions   the set of RangeActions to filter
+     * @param numberToRemove the number of RangeActions to remove from the set
      */
     private void removeRangeActionsWithBiggestImpact(Set<RangeAction> rangeActions, int numberToRemove) {
-        if (numberToRemove < 0) {
+        if (numberToRemove <= 0) {
             // Nothing to do
         } else if (numberToRemove >= rangeActions.size()) {
             rangeActions.clear();
         } else {
-            List<RangeAction> rangeActionsSortedBySensitivity = rangeActions.stream()
+            removeRangeActionsBySensitivityAndGroupId(rangeActions, numberToRemove);
+        }
+    }
+
+    private void removeRangeActionsBySensitivityAndGroupId(Set<RangeAction> rangeActions, int numberToRemove) {
+        List<RangeAction> rangeActionsSortedBySensitivity = rangeActions.stream()
                 .sorted((ra1, ra2) -> comparePrioritiesAndSensitivities(ra1, ra2, leaf.getMostLimitingElements(1).get(0), leaf))
                 .collect(Collectors.toList());
-            rangeActions.removeAll(rangeActionsSortedBySensitivity.subList(0, numberToRemove));
+
+        List<RangeAction> raHasBeenExplored = new ArrayList<>();
+        int numberToRemoveLeft = numberToRemove;
+        // Look for aligned PSTs
+        for (RangeAction ra : rangeActionsSortedBySensitivity) {
+            if (numberToRemoveLeft == 0) {
+                return;
+            }
+            // ra has already been explored.
+            if (raHasBeenExplored.contains(ra)) {
+                continue;
+            }
```

Comment:
you only need to store and test group IDs that have been explored. it will be lighter